### PR TITLE
ARROW-12859: [C++] Add ScalarFromJSON for testing

### PR DESCRIPTION
**Pull request by lidavidm**
```
ScalarFromJSON(int64(), "null") == Datum(Int64Scalar())
ScalarFromJSON(int64(), "5")    == Datum(Int64Scalar(5))
```

### CODE REVIEW

**Review by nirandaperera — APPROVED** [review]

File: cpp/src/arrow/testing/gtest_util.cc [hunk 1]
```diff
@@ -396,6 +438,13 @@ std::shared_ptr<RecordBatch> RecordBatchFromJSON(const std::shared_ptr<Schema>&
   return *RecordBatch::FromStructArray(struct_array);
 }
 
+std::shared_ptr<Scalar> ScalarFromJSON(const std::shared_ptr<DataType>& type,
+                                       util::string_view json) {
+  std::shared_ptr<Scalar> out;
+  ABORT_NOT_OK(ipc::internal::json::ScalarFromJSON(type, json, &out));
```
Comment (bkietz) [inline cpp/src/arrow/testing/gtest_util.cc:444]: This approach might be worthwhile for consistency's sake, but we could also use `Scalar::CastTo`

**Review by bkietz — APPROVED** [review]

**Comment by pitrou** [comment]
Is there anything left to do here?

**Comment by lidavidm** [comment]
> Is there anything left to do here?

I don't think so unless we really want to use Scalar::CastTo here.

**Comment by pitrou** [comment]
Can you explain the `Scalar::CastTo` suggestion? I think I'm missing something.

Also, I notice this isn't actually tested? Perhaps add a test alongside those for `ArrayFromJSON`. And ensure that a non-1 length array would return an error?

**Comment by lidavidm** [comment]
For CastTo: I think it was actually the Cast kernel (string->type cast).

I added some basic tests. I don't think there's a good way to hit the DCHECK because that would imply a JSON value converter appended two array values for a single JSON value.

**Review by pitrou — APPROVED** [review]
+1

**Pull request closed without merging**